icu4c/source/common/bytesinkutil.h

```diff
@@ -83,6 +86,42 @@ class U_COMMON_API CharStringByteSink : public ByteSink {
     CharString& dest_;
 };
 
+/**
+ * Calls a lambda that writes to a ByteSink with a CheckedArrayByteSink and
+ * then returns through u_terminateChars(), in order to implement the classic
+ * ICU4C C API writing to a fix sized buffer on top of a contemporary C++ API.
+ *
+ * @param buffer receiving buffer
+ * @param capacity capacity of receiving buffer
+ * @param status set to U_BUFFER_OVERFLOW_ERROR on overflow
+ * @param lambda that gets called with the sink as an argument
+ * @return number of bytes written, or needed (in case of overflow)
+ * @internal
+ */
+template <
+    typename F,
+    typename = std::enable_if_t<std::is_invocable_r_v<void, F, ByteSink&>>>
+int32_t
+u_withCheckedArrayByteSinkAndTerminateChars(
+        char* buffer,
+        int32_t capacity,
+        UErrorCode& status,
+        F&& lambda) {
+    if (U_FAILURE(status)) { return 0; }
+    CheckedArrayByteSink sink(buffer, capacity);
+    lambda(sink);
```

Comment:
Wouldn't it be cleaner for F to take two arguments -- both the sink and the errorCode -- rather than have the lambda definition grab the errorCode from the context?

I realize that the lambda definition already grabs other arguments from the context, but it seems extra confusing for the errorCode to be used in this wrapper and also have it bypass the wrapper for getting to the actual function inside F. Conceivably, the two UErrorCode's could be different.

Reply:
Done.